src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -906,30 +886,33 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMPing> const& m)
         fee_ = Resource::feeMediumBurdenPeer;
         m->set_type (protocol::TMPing::ptPONG);
         send (std::make_shared<Message> (*m, protocol::mtPING));
-
         return;
     }
 
-    if ((m->type () == protocol::TMPing::ptPONG) && m->has_seq ())
+    if (m->type () == protocol::TMPing::ptPONG)
     {
         // We have received a pong, update our latency estimate
-        auto unknownLatency = std::chrono::milliseconds (-1);
-
         std::lock_guard<std::mutex> sl(recentLock_);
 
-        if ((lastPingSeq_ != 0) && (m->seq () == lastPingSeq_))
+        if (m->has_seq() && lastPingSeq_ && m->seq() == lastPingSeq_)
         {
             no_ping_ = 0;
             auto estimate = std::chrono::duration_cast <std::chrono::milliseconds>
                 (clock_type::now() - lastPingTime_);
-            if (latency_ == unknownLatency)
-                latency_ = estimate;
+
+            // Calculate the cumulative moving average of the latency:
+            if (latency_)
+                latency_ = (*latency_ * 7 + estimate) / 8;
             else
-                latency_ = (latency_ * 7 + estimate) / 8;
+                latency_ = estimate;
+
+            // Don't reset the ping sequence unless we actually receive a
+            // PONG with the correct cookie. That way, peers which respond
+            // with incorrect cookies will eventually time out.
+            lastPingSeq_ = 0;
         }
         else
-            latency_ = unknownLatency;
-        lastPingSeq_ = 0;
+            latency_.reset();
```

Comment:
After reading the code more, I think that the reest should go away completely from here. This will cause the latency to remain stable until the next time we try to ping, when it will be updated by the ping handler